nameserver/zone_lookup.go

```diff
@@ -0,0 +1,342 @@
+	uniq := newUniqZoneRecords()
+	for identName, nameset := range zone.idents {
+		if identName != defaultRemoteIdent {
+			for _, ze := range lfun(nameset) {
+				uniq.add(ze)
+			}
+		}
+	}
+	if len(uniq) == 0 {
+		return nil, LookupError(target)
+	} else {
+		return uniq.toSlice(), nil
+	}
+}
+
+// Perform a lookup for a name in the zone
+// The name can be resolved locally with the local database
+func (zone *zoneDb) LookupName(name string) (res []ZoneRecord, err error) {
+	zone.mx.RLock()
+	defer zone.mx.RUnlock()
+
+	// note: LookupName() is usually called from the mDNS server, so we do not touch the name
+	name = dns.Fqdn(name)
+	Debug.Printf("[zonedb] Looking for name '%s' in local database", name)
+	return zone.lookup(name, func(ns *namesSet) []*recordEntry { return ns.getEntriesForName(name) })
+}
+
+// Perform a lookup for a IP address in the zone
+// The address can be resolved locally with the local database
+func (zone *zoneDb) LookupInaddr(inaddr string) (res []ZoneRecord, err error) {
+	zone.mx.RLock()
+	defer zone.mx.RUnlock()
+
+	// note: LookupInaddr() is usually called from the mDNS server, so we do not touch the name
+	revIPv4, err := raddrToIPv4(inaddr)
+	if err != nil {
+		return nil, newParseError("lookup address", inaddr)
+	}
+	Debug.Printf("[zonedb] Looking for address in local database: '%s' (%s)", revIPv4, inaddr)
+	return zone.lookup(inaddr, func(ns *namesSet) []*recordEntry { return ns.getEntriesForIP(revIPv4) })
+}
+
+// Perform a domain lookup with mDNS
+func (zone *zoneDb) domainLookup(target string, lfun ZoneLookupFunc) (res []ZoneRecord, err error) {
+	// no local results have been obtained in the local database: try with a mDNS query
+	Debug.Printf("[zonedb] '%s' not in local database... trying with mDNS", target)
+	lanswers, err := lfun(target)
+	if err != nil {
+		Debug.Printf("[zonedb] mDNS lookup error for '%s': %s", target, err)
+		return nil, err
+	}
+
+	// if the request has been successful, save the IP in the local database and return the corresponding ZoneRecord
+	// (we do not get the remote ident in the mDNS reply, so we save it in a "remote" ident)
+	Debug.Printf("[zonedb] adding '%s' (obtained with mDNS) to '%s'", lanswers, target)
+	res = make([]ZoneRecord, len(lanswers))
+	zone.mx.Lock()
+	now := zone.clock.Now()
+	uniq := newUniqZoneRecords()
+	for _, answer := range lanswers {
+		r, err := zone.getNamesSet(defaultRemoteIdent).addIPToName(answer, now)
+		if err != nil {
+			zone.mx.Unlock()
+			Warning.Printf("[zonedb] '%s' insertion for %s failed: %s", answer, target, err)
+			return nil, err
+		}
+		uniq.add(r)
+	}
+	zone.mx.Unlock()
+
+	return uniq.toSlice(), nil
+}
+
+// Perform a lookup for a name in the zone
+// The name can be resolved locally with the local database or with some other resolution method (eg, a mDNS query)
+func (zone *zoneDb) DomainLookupName(name string) (res []ZoneRecord, err error) {
```

Comment:
If I am reading this correctly then the very first time I perform a query on a given weaveDNS peer:
* If the name is in the local database, then I will receive only those local addresses
* If the name is not in the local database, then I will receive only addresses from the first weaveDNS peer to respond by mDNS (due to use of `zone.mdnsCli.LookupName` on line 145)

Is that right? The second case is of particular concern, because I'm likely to encounter it the very first time I try to play with load balancing - if I set up two server containers on different peers and give them both the same FQDN, then try to resolve that name from a client container on a third peer I will receive only one of the addresses the first time round. I understand there is a latency tradeoff here, but I think it would be better to use `InsistentLookupName` and put up with the slight pause on the first query...